### ISSUE TITLE
Action token signature invalid after upgrade to 22.0.x

### ISSUE THREAD

**Issue by loudmouth** [keycloak/keycloak#22447]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

token-exchange

### Describe the bug

Our authentication flow has a custom step after username/password that fetches information from another service in order to add a custom claim to user tokens. We've had a custom Keycloak extension handle the communication to the external service for a long time. The external authorization requires a valid action token from the client (Keycloak). Our extension therefore serializes an action token with the following code (some redacted for security/privacy reasons)

```java
public class OurActionTokenClass extends DefaultActionToken {
    public static final String ACTION_TYPE = "OUR_ACTION_TYPE";

    public OurActionActionToken(final String userId,
            final int absoluteExpirationInSecs) {
        super(userId, ACTION_TYPE, absoluteExpirationInSecs, null);
    }

    public String serialize(final KeycloakSession session, final RealmModel realm, final UriInfo uri) {
        return serialize(session, realm, Urls.realmIssuer(uri.getBaseUri(), realm.getName()));
    }

    public String serialize(final KeycloakSession session, final RealmModel realm, final String issuerUri) {
        final var key = session.keys().getActiveKey(realm, KeyUse.SIG, Algorithm.RS256);
        final var signatureContext = new AsymmetricSignatureSignerContext(key);

        this.iat((long) Time.currentTime()).id(getActionVerificationNonce().toString())
                .issuer(issuerUri).audience(issuerUri);

        return new JWSBuilder().kid(key.getKid()).jsonContent(this).sign(signatureContext);
    }
}
```

When upgrading from version 21.1.2 to 22.0.0 and 22.0.1, we see that while these action tokens are issued, they are issued with an invalid signature. Our Realm settings also use the RS256 signature.

We've been using Keycloak since version 3.5.4 and this has code has worked from the beginning, but it seems to break with 22.0.x. This is a blocker for upgrading for us.

### Version

22.0.1

### Expected behavior

We expect that action tokens issued and signed in a custom Keycloak extension be signed with a valid signature. 

### Actual behavior

The tokens issued in this manner have invalid signatures with Keycloak 22.0.x. 

### How to Reproduce?

The signature of After issuing a token with the code pasted above in a running instance of Keycloak with valid certificates, it's easy enough confirm the signature is invalid with jwt.io.

**Comment by rmartinc** [comment]
@loudmouth Sometimes I really wonder what you are expecting from us with an issue like this one. IMHO You have to investigate and report a proper bug saying what we have broken in the API (if something is broken) and giving an easy to test reproducer. For example in this case, giving an extension or something we can test that shows the broken API.

I have created a little extension [signature-test.zip](https://github.com/keycloak/keycloak/files/12359160/signature-test.zip) with your exact code. The extension is a realm resource. I have tested it with a default keycloak 22.0.1 (just unzipped) and started in dev mode. This way it seems that jwt.io is able to download the keys from the well known address of the iss claim.

```
unzip signature-test.zip
cd signature-test/
mvn clean package
cp target/signature-test-1.0.0-SNAPSHOT.jar /path/to/keycloak-22.0.1/providers/
cd /path/to/keycloak-22.0.1/bin
./kc.sh start-dev
```

And if I access to the URL to download the created token: http://localhost:8080/realms/master/hello

It's valid for jwt.io. Or at least it says it's valid signature.

![Screenshot from 2023-08-16 14-10-24](https://github.com/keycloak/keycloak/assets/26901232/51575149-5665-48c7-b24d-e8c1bbff682b)

Am I missing something?

**Comment by loudmouth** [comment]
Hi Ricardo,

Thanks for the quick reply, and I'm sorry for the delay in getting back to you. I do appreciate the burden of maintaining open source software as I used to do so myself, and I'm aware there are likely many issues that are not informative enough, or come from such a place of naïveté that it can be annoying. I did my best to write a clear, yet concise, issue here, but I am neither a Java developer, nor super experienced extending Keycloak. I was nominated by my team to open this issue since opening issues on open source projects can be intimidating, and I am the native English speaker/writer.

Part of the urgency in opening an issue is the understanding that the Keycloak team only supports the latest major version (currently 22), and our interpretation of that is that security patches won't be back-ported to older versions like 21. 

Anyway, we found that to get our extension working again, we had to add the following to our extension of `DefaultActionToken`

```java

    @Override
    public TokenCategory getCategory() {
        return TokenCategory.ADMIN;
    }
```

We cannot be sure why we needed to do this, but adding this override ensures that our token encoding algorithm is correct.

I suppose this issue can be closed at this point, and maybe this comment can help anybody who encounters a similar issue.

**Comment by rmartinc** [comment]
Thanks for the info @loudmouth!

The `getCategory` method is needed because there are different types of types action tokens and depending the type a different signature is used via the [TokenManager](https://github.com/keycloak/keycloak/blob/22.0.1/services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java#L190-L208). Note that using the exact code presented in the issue that method is not needed (the code doesn't use `session.tokens().encode(this)` and performs the signature directly using `JWSBuilder`). In my example I tried both ways and I added the method to make it work with the new idea.

Anyway I'm glad to hear that it's working for you. I'm closing the issue then,
